### PR TITLE
feat: add rpc generate_block_with_template to IntegrationTest rpc module

### PR DESCRIPTION
**Pull request by quake**
This PR adds `generate_block_with_template` rpc, so that dApps can get block template from `get_block_template` rpc, and then add or remove tx / proposal / uncle data in block template, and finally submit it via this rpc to control the newly generated block data.

btw, I think that we should add rpc doc to the `IntegrationTest` module, will create a new PR later.

### CODE REVIEW

File: rpc/src/module/test.rs [hunk 1]
```diff
@@ -165,4 +159,69 @@ impl IntegrationTestRpc for IntegrationTestRpcImpl {
...
+                    "resolve transactions error when generating block with block template, error: {:?}",
+                    err
+                );
+                err
+            }).ok()
+        }).collect();
+
+        let mut update_dao_template = block_template;
+        update_dao_template.dao = DaoCalculator::new(consensus, snapshot)
+            .dao_field(&rtxs, tip_header)
```
Comment (keroro520) [inline rpc/src/module/test.rs:198]: assert `block_template.parent == tip_header.hash` before this line.

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: rpc/src/module/test.rs (hunk 1 above)
Comment (quake) [reply rpc/src/module/test.rs:198]: I expect this rpc to be used to submit uncle blocks as well, so I don't want to add this condition。

Comment (keroro520) [reply rpc/src/module/test.rs:198]: `fn dao_field`'s second parameter is "parent header". And here hard-code `tip_header` as parent header, which means the block's parent is `tip_header`.

If block_template.parent != tip_header.hash, the `fn dao_field` return the wrong value.

Comment (quake) [reply rpc/src/module/test.rs]: aha, copy / paste error, now fixed, please review again, thanks.

**Review by doitian — DISMISSED** [review]

File: rpc/src/module/test.rs [hunk 2]
```diff
@@ -165,4 +159,71 @@ impl IntegrationTestRpc for IntegrationTestRpcImpl {
...
+        let transactions_provider = TransactionsProvider::new(txs.as_slice().iter());
+        let overlay_cell_provider = OverlayCellProvider::new(&transactions_provider, snapshot);
+
+        let rtxs: Vec<_> = txs.iter().filter_map(|tx| {
+            resolve_transaction(
+                tx.clone(),
+                &mut seen_inputs,
+                &overlay_cell_provider,
+                snapshot,
+            ).map_err(|err| {
```
Comment (keroro520) [inline rpc/src/module/test.rs:189]: The snapshot's alive cellset is based on `snapshot.tip` but not `parent`.
So I doubt this line has a bug.
For example, the `tx` spends an input cell that is alive for `parent` but not for `snapshot.tip`.

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: rpc/src/module/test.rs (hunk 2 above)
Comment (quake) [reply rpc/src/module/test.rs:189]: unresolvable tx will be ignored by `filter_map`, we cannot find a way to provide parent block's live cell set in current code.

Comment (keroro520) [reply rpc/src/module/test.rs:189]: I see that the current code cannot provide the alive cellset for non-tip blocks.
Then I suggest that do not implicitly filter the unresolvable transactions but just explicitly throw an error to describe why this RPC cannot accept these transactions. As a user, a filtered block is not my expected.

Comment (quake) [reply rpc/src/module/test.rs:181]: updated, please review again, thanks.

**Review by keroro520 — DISMISSED** [review]

**Comment by quake** [comment]
@doitian @keroro520 rebased with dev branch, please review again, thanks

**Comment by keroro520** [comment]
The commit message is "wip"

**Comment by quake** [comment]
> The commit message is "wip"

sorry, I forgot to squash the commits, now fixed

**Review by doitian — DISMISSED** [review]

**Review by keroro520 — APPROVED** [review]

**Comment by keroro520** [comment]
bors r=doitian,keroro520

**Review by keroro520 — APPROVED** [review]

**Review by doitian — APPROVED** [review]

**Comment by doitian** [comment]
bors retry

**Pull request merged**